corehq/apps/analytics/ab_tests.py

```diff
@@ -1,53 +1,128 @@
+        self.config = config
         self.request = request
 
     @property
     def _cookie_id(self):
-        return "{}_ab".format(self.slug)
+        return "{}_ab".format(self.config.slug)
+
+    @property
+    def _cache_id(self):
+        return "{}_{}".format(self._cookie_id, self.request.session.session_key)
 
     @memoized
     def version(self, assign_if_blank=True):
-        if self._cookie_id in self.request.COOKIES:
-            return self.request.COOKIES[self._cookie_id]
-        if assign_if_blank:
-            return random.choice(self.options)
-        return None
+        value = None
+        if cache.get(self._cache_id):
+            value = cache.get(self._cache_id)
+            self._debug_message("fetched version from cache '{}'".format(value))
+        elif (self._cookie_id in self.request.COOKIES
+              and not self.config.force_refresh):
+            value = self.request.COOKIES[self._cookie_id]
+            self._debug_message("fetched version from cookie '{}'".format(value))
+        elif assign_if_blank:
+            value = random.choice(self.config.options)
+            self._debug_message("fetched new version '{}'".format(value))
+        self.cache_version(value)
+        return value
+
+    def cache_version(self, version):
+        self._debug_message("cache version '{}' under '{}'".format(
+            version, self._cache_id))
+        cache.set(self._cache_id, version)
+        # fyi default cache timeout is 300 sec (5 min)
+
+    def _clear_cache(self):
+        self._debug_message("clearing cache '{}'".format(self._cache_id))
+        cache.delete(self._cache_id)
+
+    def _debug_message(self, message):
+        if self.config.is_debug:
+            print("SESSION AB TEST DEBUG [{}]: {}".format(self._cookie_id, message))
 
     def update_response(self, response):
-        response.set_cookie(self._cookie_id, self.version())
+        if self.config.force_refresh:
+            self._clear_response(response)
+        version = self.version()
+        self._debug_message("storing cookie value '{}' in '{}'".format(
+            version, self._cookie_id))
+        response.set_cookie(self._cookie_id, version)
+
+    def _clear_response(self, response):
+        self._debug_message("clearing cookies '{}'".format(self._cookie_id))
+        response.delete_cookie(self._cookie_id)
 
     @property
     def context(self):
-        return {
-            'name': self.name,
+        """
+        Assign this to a variable in your template context.
+
+        USAGE IN TEMPLATE:
+        {% analytics_ab_test '<api>.<slug>' <context_variable> %}
+
+        <api> - the api name (e.g. kissmetrics)
+        <slug> - is a unique slug (can be different from the slug in AbTestConfig
+        <context_variable> - the context variable you assign this property to
+
+        kissmetrics is already set up to auto detect analytics_ab_test tags and
+        send the results on page load. Please do not manually do this!
```

Comment:
What does it mean to manually detect/send? I'm not sure what to avoid doing.

Reply:
you don't have to call `identifyTraits` also, you should pass the `context` of the `SessionAbTest` directly to the template and not modify it like it is currently. cc @gbova

Reply:
Is there any code currently in prod that should be updated to reflect this? Like [new_user.ko.js](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/registration/static/registration/js/new_user.ko.js#L55) is still calling `identifyTraits`. It's not blocking this PR - and I don't think you have to do it, you can tell me or @gbova to do it - but it'd be good not to have bad examples lying around.

Reply:
will do in a separate PR since I want to make sure this PR doesn't mess up an existing test and ends up needing a revert

Reply:
good call